test/e2e/kubelet.go

```diff
@@ -127,8 +131,142 @@ func updateNodeLabels(c clientset.Interface, nodeNames sets.String, toAdd, toRem
+			APIVersion: registered.GroupOrDie(v1.GroupName).GroupVersion.String(),
+		},
+		ObjectMeta: v1.ObjectMeta{
+			GenerateName: "pod-nfs-vol-",
+			Namespace:    ns,
+		},
+		Spec: v1.PodSpec{
+			Containers: []v1.Container{
+				{
+					Name:    "pod-nfs-vol",
+					Image:   "gcr.io/google_containers/busybox:1.24",
+					Command: []string{"/bin/sh"},
+					//Note: timing window where nfs-server pod is deleted while/before this pod
+					//  accesses the nfs mount. In this case this pod won't be deleted regardless
+					//  of the wait time. This scenario should be its own It test.
+					///Args:    []string{"-c", "rm /mnt/SUCCESS* && touch /mnt/SUCCESS && cat /mnt/SUCCESS && sleep 6000"},
+					Args: []string{"-c", "sleep 6000"},
+					VolumeMounts: []v1.VolumeMount{
+						{
+							Name:      "nfs-vol",
+							MountPath: "/mnt",
+						},
+					},
+					SecurityContext: &v1.SecurityContext{
+						Privileged: &isPrivileged,
+					},
+				},
+			},
+			RestartPolicy: v1.RestartPolicyOnFailure,
+			Volumes: []v1.Volume{
+				{
+					Name: "nfs-vol",
+					VolumeSource: v1.VolumeSource{
+						NFS: &v1.NFSVolumeSource{
+							Server:   ip,
+							Path:     "/exports",
+							ReadOnly: false,
+						},
+					},
+				},
+			},
+		},
+	}
+	rtnPod, err := c.Core().Pods(ns).Create(pod)
+	Expect(err).NotTo(HaveOccurred())
+	Expect(rtnPod).NotTo(BeNil())
+
+	// wait for pod to be ready
+	err = f.WaitForPodRunningSlow(rtnPod.Name) // longer wait
+	Expect(err).NotTo(HaveOccurred())
+
+	return rtnPod
+}
+
+// Deletes the passed-in pod and conditionally waits for the pod to be terminated.
+func deletePodwithWait(f *framework.Framework, c clientset.Interface, pod *v1.Pod, wait bool) {
+
+	framework.Logf("Deleting pod %v", pod.Name)
+	err := c.Core().Pods(pod.Namespace).Delete(pod.Name, nil)
+	Expect(err).NotTo(HaveOccurred())
+
+	if wait {
+		// wait for pod to terminate. Expect apierr NotFound
+		err = f.WaitForPodTerminated(pod.Name, "")
+		Expect(err).To(HaveOccurred())
+		if !apierrs.IsNotFound(err) {
+			framework.Failf("Unexpected error deleting pod %v: %v", pod.Name, err)
+		}
+		framework.Logf("Pod %v successfully deleted", pod.Name)
+	}
+}
+
+// Checks for a lingering nfs mount and/or uid directory on the passed-in pod's node *after*
+// pod has been deleted. If expectClean is true then we expect the node to be cleaned up and
+// thus commands like `ls <uid-dir>` should fail (since that dir was removed). If expectClean
+// is false then we expect the node is not cleaned up, and thus cmds like `ls <uid-dir>`
+// should succeed.
+// Note: uses ssh to the target node.
+// Note: expects framework.TestContext.Provider to be filled in.
+// Note: since the passed-in pod has been deleted the pod api object is stale.
+func checkPodCleanup(pod *v1.Pod, expectClean bool) {
+
+	check := func(e int) {
+		if expectClean {
+			Expect(e).NotTo(BeZero())
+		} else {
+			Expect(e).To(BeZero())
+		}
+	}
+
+	nodeName := pod.Spec.NodeName
+	if len(nodeName) == 0 {
```

Comment:
When deployed locally both _pod.Spec.NodeName_ and _pod.Status.PodUID_ are empty. Is this expected? Is there another way to get the node name of the target pod's host? 
Note: `framework.SSH()` seems to work when the node name is empty.